### PR TITLE
chore: More robust caching strategy for pnpm installs in circle ci

### PR DESCRIPTION
**Pull request by roninjin10**
Changes our caching strategy to only cache based on pnpm-lockfile.yaml instead of packages/*/package.json. This has 2 advantages:
- Removes false cache misses when the package.json changes but the deps didn't
- Makes the circle config more robust to breakage. We don't have to manually keep it in sync everytime we add and remove a package
  - This protects us since the first time we add a new package it will work and pass checks and possibly merge if we forget and then fail on second run when cache is badly populated

### CODE REVIEW

**Comment by roninjin10** [comment]
Current dependencies on/for this PR:
* develop
  * **PR #6287** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/optimism/6287" target="_blank">Graphite</a> 
    * **PR #6210** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/optimism/6210" target="_blank">Graphite</a> 
      * **PR #6303** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/optimism/6303" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/ethereum-optimism/optimism/6303?utm_source=stack-comment).

**Comment by roninjin10** [comment]
CI won't actually pass until ci builder is updated to have yarn

**Review by mslipper — APPROVED** [review]

**Review by tynes — APPROVED** [review]

**Comment by roninjin10** [comment]
This looks good to merge since CI passed

**Comment by roninjin10** [comment]
It's stacked on my own pr though so don't merge it til that one merges

**Review by tynes — CHANGES_REQUESTED** [review]
This has merge conflicts in it now

**Comment by tynes** [comment]
@roninjin10 I just rebased this

**Comment by roninjin10** [comment]
@tynes oops I likely just force pushed over your changes

I noticed a bunch of unnecessary stuff in the config. Will likely reset to your rebase @tynes and then open a new pr for that. e.g. stuff that doesn't depend on node_modules was still restoring cache etc.

Don't merge this til I remove the save-cache. The save-cache should only exist in the monorepo job

File: .circleci/config.yml
Comment (roninjin10) [inline .circleci/config.yml:120]: This is unecessary pretty sure actually could ust delete

File: .circleci/config.yml
Comment (roninjin10) [inline .circleci/config.yml:131]: these can go away and then anything that needs node modules can just `pnpm i --offline-only` to instantly install from cache

Comment (roninjin10) [reply .circleci/config.yml:131]: will move to it's own pr though

File: .circleci/config.yml
Comment (roninjin10) [inline .circleci/config.yml]: --prefer-offline is faster if there is any cache

Comment (roninjin10) [reply .circleci/config.yml]: again will move to it's own pr

**Comment by tynes** [comment]
Lemme know when this is ready for review again

**Comment by roninjin10** [comment]
I think someone got to this before I got back?

**Comment by roninjin10** [comment]
will do some of the other cleanups I added

**Comment by roninjin10** [comment]
Dismissed @mslipper review since this is a different pr now
@tynes I'm likely adding a `pnpm i --prefer-offline` in places it isn't necessary and relying on you to point them out

File: .circleci/config.yml
Comment (roninjin10) [reply .circleci/config.yml:131]: This own pr is now this pr

Comment (roninjin10) [reply .circleci/config.yml:131]: Doing --prefer-offline instead of --offline-only just because it makes it less fragile to breaking with the same upside

**Review by mslipper — APPROVED** [review]

File: .circleci/config.yml
Comment (tynes) [inline .circleci/config.yml]: I don't think we need JS deps in this task

File: .circleci/config.yml
Comment (tynes) [inline .circleci/config.yml]: We def don't need JS deps in this task

File: .circleci/config.yml
Comment (tynes) [inline .circleci/config.yml]: We need JS deps in this task but will not need them in the future

File: .circleci/config.yml
Comment (tynes) [inline .circleci/config.yml]: We don't need JS deps in this task

File: .circleci/config.yml
Comment (tynes) [inline .circleci/config.yml]: Don't need JS deps, tbh i think we can remove this entire task

File: .circleci/config.yml
Comment (tynes) [inline .circleci/config.yml]: Same, don't need JS deps, can prob remove the entire task

File: .circleci/config.yml
Comment (tynes) [inline .circleci/config.yml]: Don't need JS

File: .circleci/config.yml
Comment (tynes) [inline .circleci/config.yml]: We now install solidity deps with git submodules so I don't think we need yarn here unless for a diff reason from the removed comment

File: .circleci/config.yml
Comment (tynes) [inline .circleci/config.yml]: No need for JS deps here

**Review by tynes — CHANGES_REQUESTED** [review]
A lot of additions that we don't need, would prefer to keep as much JS out of the contracts and Go CI as possible due to it not being necessary

**Comment by roninjin10** [comment]
Will get back to this this week

File: .circleci/config.yml
Comment (roninjin10) [reply .circleci/config.yml]: Oh yea likely fixed now

**Comment by roninjin10** [comment]
This pr blocks @spacesailor24 pr https://github.com/ethereum-optimism/optimism/pull/6619

File: .circleci/config.yml
Comment (tynes) [reply .circleci/config.yml]: Confused why this is here, wont this slow down things? no js is needed here

**Review by tynes — APPROVED** [review]

**Pull request merged**
